app/discovery/provider/consulcatalog/consulcatalog.go

```diff
@@ -0,0 +1,95 @@
+package consulcatalog
+
+import (
+	"context"
+	"fmt"
+	"github.com/umputun/reproxy/app/discovery"
+	"log"
+	"sync"
+	"time"
+)
+
+//go:generate moq -out consul_client_mock.go -skip-ensure -fmt goimports . ConsulClient
+
+type ConsulClient interface {
+	Get() ([]*ConsulService, error)
+}
+
+type ConsulCatalog struct {
+	client  ConsulClient
+	events  chan discovery.ProviderID
+	autoAPI bool
+
+	listMx sync.RWMutex
+	list   []*ConsulService
+
+	currentServicesID map[string]struct{}
+}
+
+func New(ctx context.Context, client ConsulClient, autoAPI bool, updateInterval time.Duration) *ConsulCatalog {
+	cc := &ConsulCatalog{
+		client:            client,
+		autoAPI:           autoAPI,
+		events:            make(chan discovery.ProviderID),
```

Comment:
those events relevant for Events func only and better be local. See how this done in Docker provider